Merlin the wizard is testing a new magic potion to transform ferocious hyenas into cute dogs. The powerful concoction must contain a minimum of 15 grams of loyalty, 21 g of affection, and 2 g of obedience. However, the dogs should be fed no more than 34 grams of potion per day.

Rather than making a new potion from scratch, Merlin will reuse the potions X and Y and blend them for an optimal mix. Potion X contains 6 g of loyalty, 14 g of affection, and 1 g of obedience, and costs € 2 per gram. Potion Y contains 11 g of loyalty, 30 g of affection, and 2 g of obedience, at a cost of € 4 per gram.

What is the optimal blend for the new magic potion?

In [25]:
from ortools.linear_solver import pywraplp

In [26]:
solver = pywraplp.Solver.CreateSolver('GLOP')

In [27]:
x = solver.NumVar(0, solver.infinity(), 'x')
y = solver.NumVar(0, solver.infinity(), 'y')

In [28]:
solver.Add(x * 6 + y * 11 >= 15) 

solver.Add(x * 14 + y * 30 >= 21)

solver.Add(x * 1 + y * 2 >= 2)

solver.Add(x + y <= 34)

<ortools.linear_solver.pywraplp.Constraint; proxy of <Swig Object of type 'operations_research::MPConstraint *' at 0x7fb500124960> >

In [29]:
solver.Maximize(2 * x + 4 * y)

In [30]:
status = solver.Solve()

In [31]:
if status == pywraplp.Solver.OPTIMAL:
    print('Solution:')
    print('Objective value =', solver.Objective().Value())
    print('x =', x.solution_value())
    print('y =', y.solution_value())
else:
    print('The problem does not have an optimal solution.')

Solution:
Objective value = 136.00000000000003
x = 0.0
y = 34.00000000000001


The optimal blend for the potion appears to be 0 grams of X and 34 grams of Y at a cost of 136 €